Enable IRQ when binding to a notification

Indanz opened the pull request:
And disable when unbinding.

This solves issue #1536.

Indanz committed:
Enable IRQ when binding to a notification

And disable when unbinding.

This solves issue #1536.

Signed-off-by: Indan Zupancic <indan@nul.nu>

midnightveil reviewed (approved):
Oh, that's nice. It also seems to have removed some complexity with the delete/deletingIRQHandler too, which is great. I love it when this sort of improvement makes the code nicer.

Indanz commented:
> Oh, that's nice. It also seems to have removed some complexity with the delete/deletingIRQHandler too, which is great. I love it when this sort of improvement makes the code nicer.

I like that too, but when the obvious solution is better and simpler than what's actually implemented, it makes me highly suspicious that there is some subtle or obscure reason for the current implementation to be as it is.

lsf37 commented:
Hm, yes, this is suspiciously simple. I'll see if I can find something in the earlier commit history that suggests why we did not do this in the first place.

lsf37 commented on src/object/interrupt.c:
```diff
@@ -178,12 +179,14 @@ void invokeIRQHandler_ClearIRQHandler(irq_t irq)
+
     irqSlot = intStateIRQNode + IRQT_TO_IDX(irq);
     /** GHOSTUPD: "(True, gs_set_assn cteDeleteOne_'proc (-1))" */
     cteDeleteOne(irqSlot);
```
This breaks the invariant that when an IRQHandlerCap exists in an IRQ slot, the corresponding IRQ must be enabled. `cteDelete` and its variations, including the simpler `cteDeleteOne` are not functions where it would be easy to deal with temporarily broken invariants. If it is acceptable to do the `setIRQState` after `cteDeleteOne`, we might be Ok.

midnightveil replied:
That invariant being broken is somewhat the point here, it would need to become "IRQ enabled whenever irq handler and bound notification".

Indanz replied:
I don't think it's possible for the final cap to be in the IRQ CNode, that's always a derived cap, never the final, I think. Does that help?

lsf37 replied:
> I don't think it's possible for the final cap to be in the IRQ CNode, that's always a derived cap, never the final, I think. Does that help?

It is absolutely possible for the final cap to be in the IRQ CNode, you just need to delete the parent cap.

Indanz replied:
But wouldn't that also delete/revoke the derived caps? Or is that a special case of some cap types? Or am I just confused and it never works like that for any caps?

Indanz replied:
(You can keep MCS O(1) with careful design, also for the release queue.)

lsf37 replied:
From user level, sure, but it is no longer guaranteed by the kernel (where it was before). I don't like kernel properties that depend on well-behaving users. That's not really a kernel then, that's just a library. It's not quite as black and white as that, I know. Still, I don't like it.

midnightveil commented on src/object/interrupt.c:
```diff
@@ -192,11 +195,6 @@ void deletingIRQHandler(irq_t irq)
     cteDeleteOne(slot);
 }
```
From the perspective of easing the proof burden (since I doubt you're being paid for this) and a (extremely small, like maxIrq/64 words) memory overhead, would it beasier to just use 2 separate tables and make an issue for combining them? Or should I just try doing a combined one and seeing how it goes?

From a software side it's somewhat easy either way...

lsf37 replied:
I'd go with a combined one. If I'm spending free time on it, I'd rather only do it once.

Indanz commented:
Thanks for the update.

> For acknowledging an interrupt, we currently require that `interrupt_states s irq != IRQInactive`. With the new invariant we no longer know this, and it can in fact be violated, because acknowledging an interrupt only requires the `IRQHandler` cap. If you call Ack before setting a notification, the condition will be violated, because the presence of that cap alone no longer guarantees that the interrupt is not inactive.
> 
> We could fix this by checking for the presence of a notification cap in the IRQ node, because if we have the handler cap, we can locate the slot.

Seems easier to check for `interrupt_states s irq != IRQInactive` directly.

> This would either create a new error condition for the user or would be silently skipped. I think it should create an error, because if you're acknowledging interrupts before you can receive them, something is probably wrong.

I don't think so, the end state is as the user wants it to be, so I wouldn't return an error. Currently the call doesn't return an error when there is nothing to ack on a hw level either, this wouldn't be much different. Edit: Not exactly the same, there probably will be a notification set at least when the user acks.

Also acking first is currently needed sometimes to avoid the problem mentioned in the issue, all that code would trigger an error.

> The other occurrences so far point at a bit of trouble with the way we will need to phrase the new invariant. Currently, deleting caps does not violate the invariant, because it only says "if handler cap exists, then X". So if you remove a handler cap, that remains trivially true. If we now say, "if no handler cap exists, then Y", we will need to show that corresponding cleanup has happened already. But maybe we don't need that direction, and only need the presence of the notification.

Well, if that direction isn't trivial, I wouldn't bother with it.

> Tying interrupt state to the presence of notifications will make a bunch of things harder, because we will now need to know where caps are when we do any cap operation, but I think we have most of those conditions already present for normal cases of copy/move/swap/mint/mutate.

The IRQ notification cap can't be modified via those operations, is that what you mean?

> cteInsert will gain another special case, although we can hopefully remove or at least reduce the current special case for IRQHandler caps (maybe).

If it's easier, what about only proving that the IRQ gets enabled after a `SetIRQHandler` call (and disabled after `ClearIRQHandler`), and not tying it to the cap state at all? Because with the current version the IRQ doesn't get disabled when the cap disappears, so no need to track that in the proofs either.

Indanz committed:
ACK irq in seL4_IRQHandler_SetNotification

When a notification is revoked, the IRQ won't be disabled, that
only happens after seL4_IRQHandler_Clear(). An interrupts after
this will need to be ACked, but user space never received it.

Similarly, after user space received an IRQ via a notification,
if the task handling the IRQ gets destroyed or restarted before
it had a chance to ACK the interrupt, the new task also won't
know that an ACK is still needed.

With this change user space will always receive interrupts after
doing an seL4_IRQHandler_SetNotification() call, no matter what
happened before.

Signed-off-by: Indan Zupancic <indan@nul.nu>

Indanz committed:
Merge branch 'master' into irq_issue_1536
